### PR TITLE
StagedTuple changes 

### PR DESCRIPTION
**Pull request by biboudis**
Restore previous `unapply` and homogenize staged method names. I used the convention of putting the `*Staged` at the end. 

- Decouple staged tuples from dynamic tuples 
- Add an explicit argument `[Tail[Tup]]` because `tail` (on a Tuple6) was failing even on master
- Refactor direct usages of throw `QuoteError`

depends on #6210, checked with `stageIt = true` that it passes and reverted it back to `false`

### CODE REVIEW

**Comment by biboudis** [comment]
Rebased 

Note to self: an assertion error occurs after erasure. It seems the following synthesised code raises an error:

```scala
.asInstanceOf[{
  type evidence$22$_~$15 = Int
  type evidence$21$_~$5 = (Int, Int)(Tuple_this)(Tuple_this)
  evidence$22$_~$15 *: evidence$21$_~$5
}
```
instead it should have been (Int, Int, Int)

**Comment by biboudis** [comment]
Rebased and waiting for #6210. After that gets fixed do we still want to merge this? It essentially refactors StagedTuples according to #5409 but if StagedTuples are dropped ... 🤷‍♂️

ping @odersky

File: compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala [hunk 1]
```diff
@@ -45,8 +45,11 @@ class PlainPrinter(_ctx: Context) extends Printer {
   protected def debugPos: Boolean = ctx.settings.YdebugPos.value
 
   def homogenize(tp: Type): Type =
-    if (homogenizedView)
-      tp match {
+    if (homogenizedView) {
+       val a = tp.tryNormalize
+          if (a.exists) {
+            homogenize(a)
+          } else tp match {
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala]: This code needs a bit of cleanup

File: library/src-bootstrapped/scala/Tuple.scala [hunk 2]
```diff
@@ -428,53 +336,15 @@ sealed trait NonEmptyTuple extends Tuple {
...
+      'this, constValueOpt[Size[this.type]], 'n, constValueOpt[n.type])}
 }
 
 @showAsInfix
 sealed class *:[+H, +T <: Tuple] extends NonEmptyTuple
 
 object *: {
-  inline def unapply[H, T <: Tuple](x: H *: T) = (x.head, x.tail)
+  inline def unapply[H, T <: Tuple](x: H *: T) =
+    (scala.runtime.DynamicTuple.dynamicHead(x), scala.runtime.DynamicTuple.dynamicTail(x))
```
Comment (nicolasstucki) [inline library/src-bootstrapped/scala/Tuple.scala]: This change is only for staging, it should be
```scala
if (stageIt) (scala.runtime.DynamicTuple.dynamicHead(x), scala.runtime.DynamicTuple.dynamicTail(x))
else (x.head, x.tail)
```

File: library/src-bootstrapped/scala/compiletime/StagedTuple.scala [hunk 3]
```diff
@@ -1,18 +1,22 @@
-package scala
+package scala.compiletime
```
Comment (nicolasstucki) [inline library/src-bootstrapped/scala/compiletime/StagedTuple.scala]: We should move this to `scala.internal`

File: library/src-bootstrapped/scala/Tuple.scala (hunk 2 above)
Comment (biboudis) [reply library/src-bootstrapped/scala/Tuple.scala]: Should be the other way around, no? `if (!stageIt)`

**Comment by biboudis** [comment]
After rebasing and turning `stageIt` on I get:

```scala
[error] undefined: new dotty.tools.dotc.ast.Trees.Instance[dotty.tools.dotc.core.Types.Type] # 19433572: TermRef(AppliedType(TypeRef(TermRef(TermRef(ThisType(TypeRef(NoPrefix,module class dotc)),module ast),Trees),class Instance),List(TypeRef(TermRef(TermRef(ThisType(TypeRef(NoPrefix,module class dotc)),module core),Types),Type))),<init>) at frontend
dotty.tools.FatalError: 
        at dotty.tools.dotc.typer.ErrorReporting$Errors.takesNoParamsStr(ErrorReporting.scala:84)
        at dotty.tools.dotc.typer.TypeAssigner.$anonfun$assignType$8(TypeAssigner.scala:394)
        at dotty.tools.dotc.core.Types$ErrorType$$anon$8.msg(Types.scala:4172)
        at dotty.tools.dotc.printing.PlainPrinter.$anonfun$toText$1(PlainPrinter.scala:179)
        at dotty.tools.dotc.printing.PlainPrinter.controlled(PlainPrinter.scala:28)
        at dotty.tools.dotc.printing.PlainPrinter.toText(PlainPrinter.scala:131)
        at dotty.tools.dotc.printing.RefinedPrinter.super$toText(RefinedPrinter.scala:223)
        at dotty.tools.dotc.printing.RefinedPrinter.$anonfun$toText$1(RefinedPrinter.scala:242)
        at dotty.tools.dotc.printing.PlainPrinter.controlled(PlainPrinter.scala:28)
        at dotty.tools.dotc.printing.RefinedPrinter.toText(RefinedPrinter.scala:128)
        at dotty.tools.dotc.core.Types$Type.toText(Types.scala:1506)
        at dotty.tools.dotc.printing.Showable.show(Showable.scala:23)
        at dotty.tools.dotc.printing.Showable.show$(Showable.scala:23)
        at dotty.tools.dotc.core.Types$Type.show(Types.scala:87)
...
```

```
exception occurred while typechecking /Users/bibou/Projects/dotty/compiler/src/dotty/tools/dotc/ast/Desugar.scala
```

File: library/src-bootstrapped/scala/compiletime/StagedTuple.scala (hunk 3 above)
Comment (biboudis) [reply library/src-bootstrapped/scala/compiletime/StagedTuple.scala]: I tried the change and I get:

```scala
[info] Compiling 5 Scala sources to /Users/bibou/Projects/dotty/library/../out/bootstrap/dotty-library-bootstrapped/scala-0.14/classes ...
[error] -- Error: /Users/bibou/Projects/dotty/library/src-bootstrapped/scala/Tuple.scala:350:24 
[error] 350 |    if (Tuple.stageIt) (x.head, x.tail)
[error]     |                        ^^^^^^
[error]     |Could not find class scala.internal.StagedTuple$ in classpath. The most common reason for that is that you apply macros in the compilation run that defines them
[error]     | This location is in code that was inlined at Tuple.scala:224
[error]     | This location is in code that was inlined at Tuple.scala:350
[error] -- Error: /Users/bibou/Projects/dotty/library/src-bootstrapped/scala/internal/quoted/Matcher.scala:287:44 
[error] 287 |      case (Some(acc), Some(holes)) => Some(acc ++ holes)
[error]     |                                            ^^^^^^^^^^^^
[error]     |Could not find class scala.internal.StagedTuple$ in classpath. The most common reason for that is that you apply macros in the compilation run that defines them
[error]     | This location is in code that was inlined at Tuple.scala:66
[error]     | This location is in code that was inlined at Matcher.scala:287
[error] two errors found
[error] (dotty-library-bootstrapped / Compile / compileIncremental) Compilation failed
[error] Total time: 2 s, completed Apr 9, 2019 10:15:43 AM
```

**Comment by biboudis** [comment]
Can I merge this @nicolasstucki ?

**Pull request merged**